### PR TITLE
Introduce "cell splitter" script

### PR DESCRIPTION
**Pull request by jugglinmike**
Inspired by a feature request on Airtable's Community forums:

[1] https://community.airtable.com/t/split-text-to-columns/28638

### CODE REVIEW

File: cell-splitter.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+ *    of the split values.
+ */
+const tableName = 'Split Demo';
+const sourceName = 'My Things';
+const destinationNames = ['Thing 1', 'Thing 2', 'Thing 3'];
+
+// You shouldn't need to modify the code following this line to use this
+// script.
+
+async function main() {
```
Comment (s2tephen) [inline cell-splitter.js:26]: let's just inline this whole function

File: cell-splitter.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+ *    the table in your Base
+ * 2. Replace the value of the `sourceName` constant with the field which
+ *    contains the source value (that is: the value to be split).
+ * 3. Replace the value of the `destinationNames` constant with a JavaScript
+ *    array of string values, each describing the name of a field to store one
+ *    of the split values.
+ */
+const tableName = 'Split Demo';
+const sourceName = 'My Things';
+const destinationNames = ['Thing 1', 'Thing 2', 'Thing 3'];
```
Comment (s2tephen) [inline cell-splitter.js:21]: the eslint setup currently allows it, but let's use `let` instead of `const` here (the rationale is that it will be more readable to non-programmers in a "high school algebra/physics" sort of way)

File: cell-splitter.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
+/**
+ * Airtable Field Split Script
```
Comment (s2tephen) [inline cell-splitter.js:2]: nit: our copy style guide prefers sentence case, e.g. `Airtable field split script`

File: cell-splitter.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * For each record in a given Airtable Base, read the value of one cell,
```
Comment (s2tephen) [inline cell-splitter.js:4]: nit: lowercase `base`, we don't capitalize the names of concepts in Airtable

File: cell-splitter.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ * How to adapt this script
+ *
+ * 1. Replace the value assigned to the `tableName` constant with the name of
+ *    the table in your Base
+ * 2. Replace the value of the `sourceName` constant with the field which
+ *    contains the source value (that is: the value to be split).
+ * 3. Replace the value of the `destinationNames` constant with a JavaScript
+ *    array of string values, each describing the name of a field to store one
+ *    of the split values.
+ */
```
Comment (s2tephen) [inline cell-splitter.js:18]: great idea!

File: cell-splitter.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+ * Airtable Field Split Script
+ *
+ * For each record in a given Airtable Base, read the value of one cell,
+ * interpret that value as a space-separated list (e.g. the value "a b c"
+ * becomes "a", "b", and "c"), and store each value in a corresponding
+ * destination cell.
+ *
+ * How to adapt this script
+ *
+ * 1. Replace the value assigned to the `tableName` constant with the name of
```
Comment (s2tephen) [inline cell-splitter.js:11]: nit: use the word `variable` instead of `constant` (see below)

File: cell-splitter.js (hunk 2 above)
Comment (s2tephen) [inline cell-splitter.js:21]: let's add the word `Field` to these variable names, e.g. `sourceFieldName`, `destinationFieldNames`

File: cell-splitter.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+const tableName = 'Split Demo';
+const sourceName = 'My Things';
+const destinationNames = ['Thing 1', 'Thing 2', 'Thing 3'];
+
+// You shouldn't need to modify the code following this line to use this
+// script.
+
+async function main() {
+    if (typeof tableName !== 'string') {
+        throw new Error('The `tableName` constant must be a JavaScript string value.');
```
Comment (s2tephen) [inline cell-splitter.js:28]: nit: `variable` instead of `constant`

we can also shorten it to `must be a string`

File: cell-splitter.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    if (!Array.isArray(destinationNames)) {
+        throw new Error('The `destinationNames` constant must be a JavaScript array value.');
+    }
+    if (!destinationNames.length) {
+        throw new Error('The `destinationNames` array must have at least one element.');
+    }
+    if (!destinationNames.every((name) => typeof name === 'string')) {
+        throw new Error(
+            'Every element of the `destinationNames` array must be a JavaScript string value'
```
Comment (s2tephen) [inline cell-splitter.js:41]: nit: `element in the`

File: cell-splitter.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        );
+    }
+
+    let table = base.getTable(tableName);
+    let result = await table.selectRecordsAsync();
```
Comment (s2tephen) [inline cell-splitter.js:46]: nit: rename to `query`

File: cell-splitter.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+    for (let record of result.records) {
+        let values = (record.getCellValueAsString(sourceName) || '').split(/\s+/);
+
+        if (values.length > destinationNames.length) {
+            output.markdown(
+                `**WARNING**: encountered ${values.length} values, but only ` +
+                    `${destinationNames.length} destination fields are available. ` +
+                    `(values: ${values})`
```
Comment (s2tephen) [inline cell-splitter.js:55]: let's reword this to be more end user friendly. maybe something like:
```
output.markdown(`
**${record.name}** was not updated. It has more values in the ${sourceFieldName} field than there are destination fields.
`);
```

alternatively, did we consider dropping the leftover values? that might cut down on complexity

File: cell-splitter.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            );
+
+            values.length = destinationNames.length;
```
Comment (s2tephen) [inline cell-splitter.js:58]: this could use a comment

File: cell-splitter.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        //
+        //     {
+        //       "first": "foo",
+        //       "second": "bar",
+        //       "third": ""
+        //     }
+        let toUpdate = destinationNames.reduce((memo, name) => {
+            memo[name] = values.shift();
+            return memo;
+        }, {});
```
Comment (s2tephen) [inline cell-splitter.js:78]: rather than using `reduce`, let's construct an object and populate it with a `for` loop:
```
let fields = {};
for (let i = 0; i < destinationFieldNames.length; i++) {
    let fieldName = destinationFieldNames[i];
    fields[fieldName] = values[i];
}
```

File: cell-splitter.js [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            output.markdown(
+                `**WARNING**: encountered ${values.length} values, but only ` +
+                    `${destinationNames.length} destination fields are available. ` +
+                    `(values: ${values})`
+            );
+
+            values.length = destinationNames.length;
+        }
+
+        // Create an object suitable for use in `updateRecordAsync`. It keys
```
Comment (s2tephen) [inline cell-splitter.js:61]: typo: `Its`

File: cell-splitter.js [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        //       "first": "foo",
+        //       "second": "bar",
+        //       "third": ""
+        //     }
+        let toUpdate = destinationNames.reduce((memo, name) => {
+            memo[name] = values.shift();
+            return memo;
+        }, {});
+
+        await table.updateRecordAsync(record, toUpdate);
```
Comment (s2tephen) [inline cell-splitter.js:80]: we may want to use the `updateRecordsAsync` batch update method instead, since you might run this script on a bunch of records

File: cell-splitter.js [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    }
+}
+
+try {
+    await main();
+} catch (error) {
+    output.markdown(`**ERROR**: ${error}`);
+}
```
Comment (s2tephen) [inline cell-splitter.js:88]: we shouldn't need to wrap this in a try/catch, the scripting block should already render errors in a nicely formatted way

File: cell-splitter.js (hunk 10 above)
Comment (jugglinmike) [reply cell-splitter.js:55]: That's the intent of the statement `values.length = destinationFieldNames.length;`. I'll take this as further evidence of the need of a comment for that line :)

File: cell-splitter.js (hunk 14 above)
Comment (jugglinmike) [reply cell-splitter.js:80]: This script runs against all the records, and that may exceed the method's 50-record limit. The batching logic won't be too complex, but it won't be trivial, either. Since some of your other feedback relates to code complexity, I want to verify: would you like to see record batching logic here?

Comment (jugglinmike) [reply cell-splitter.js]: Scratch that. I just saw batching in [a script you wrote today](https://community.airtable.com/t/url-to-attachment/28569/2), so I've implemented the same here.

File: cell-splitter.js [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    // Create an object suitable for use in `updateRecordAsync`. Its keys
+    // are the names of the "destination" fields as specified at the
+    // beginning of this scripts, and its values are the corresponding
+    // values from the "source" field.
+    //
+    // For example, given a "source" field with the text "foo bar" and a
+    // set of "destination" fields named "first", "second", and "third",
+    // `toUpdate` will be the following JavaScript object:
```
Comment (s2tephen) [inline cell-splitter.js:67]: `toUpdate` => `fields`

File: cell-splitter.js [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+let updates = [];
+
+for (let record of query.records) {
+    let values = (record.getCellValueAsString(sourceFieldName) || '').split(/\s+/);
+
+    if (values.length > destinationFieldNames.length) {
+        output.markdown(
+            `**WARNING**: encountered ${values.length} values, but only ` +
+                `${destinationFieldNames.length} destination fields are available. ` +
+                `(values: ${values})`
```
Comment (s2tephen) [inline cell-splitter.js:53]: bumping my original copyedit suggestion:
```
output.markdown(`
**${record.name}** was not updated. It has more values in the ${sourceFieldName} field than there are destination fields.
`);
```

File: cell-splitter.js [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        );
+
+        // Remove the extraneous values from the array
```
Comment (s2tephen) [inline cell-splitter.js:56]: this also "fills" the array when there are fewer values than destination fields, correct? let's update the comment to reflect that

File: cell-splitter.js (hunk 17 above)
Comment (jugglinmike) [reply cell-splitter.js:53]: It sounds like there may be a misunderstanding about the condition this branch describes and how we handle it. Here's my interpretation.

The following condition is only true when there are more space-separated values than there are cells in which to store them:

    if (values.length > destinationFieldNames.length) {

So inside the branch, we're only concerned with what to do with "extra" values. (If there are fewer values than expected, the branch is skipped, and the cells which are unaccounted for are not modified.) I chose to discard the extra values and warn the user that this occurred.

    output.markdown(
        `**WARNING**: encountered ${values.length} values, but only ` +
            `${destinationFieldNames.length} destination fields are available. ` +
            `(values: ${values})`
    );

    // Remove the extraneous values from the array
    values.length = destinationFieldNames.length;

Your suggestion for the warning message describes a different response to this situation: advance to the next record without making any changes. I'm happy to change the logic accordingly, but I want to verify with you first. Is that the behavior you'd prefer to see?

**Comment by jugglinmike** [comment]
@s2tephen [This forum topic](https://community.airtable.com/t/best-practices-for-making-a-good-open-source-script/27629) prefers the use of the dedicated `input` methods for table names and field names:

> - Use `input.tableAsync`, `input.viewAsync`, `input.fieldAsync` and `input.recordAsync` to avoid references to specific fields or tables that might be different for other users. If you do reference specific fields or tables directly in your script, we recommend pulling them out into variables and moving them to the top of the script so others can change them to match their own bases.

Do you have any thoughts on how to choose between the two approaches?

File: cell-splitter.js (hunk 17 above)
Comment (s2tephen) [reply cell-splitter.js:53]: ah, i see — thanks for clarifying! i think the behavior you have is fine, my concern was more around the copy. i actually think we might not need to show this warning at all

**Comment by s2tephen** [comment]
@jugglinmike

> Do you have any thoughts on how to choose between the two approaches?

I added this to the styleguide, let me know if that makes sense:
> If a script is likely to use the same models every time, hardcode model names at the top of the script. Otherwise, use `input.table/view/fieldAsync`.

**Comment by jugglinmike** [comment]
Thanks again, @s2tephen!
